### ISSUE TITLE
BatchReceivePolicy.maxNumMessages not working

### ISSUE THREAD

**Issue by Myron5787** [fsharplang-ru/pulsar-client-dotnet#259]
Dear Author,

I am currently utilizing the BatchReceiveAsync method in my project, and I've encountered an issue where the number of messages received exceeds the limit set by BatchReceivePolicy.maxNumMessages. Despite configuring the BatchReceivePolicy to limit the number of messages per batch, the method occasionally returns more messages than specified.

```
await using var consumer =
    await pulsarClient.NewConsumer(Schema.STRING(Encoding.UTF8))
        .Topic(_topic)
        .SubscriptionName(_subscription)
        .SubscriptionType(SubscriptionType.Shared)
        .SubscriptionInitialPosition(SubscriptionInitialPosition.Latest)
        .AutoUpdatePartitions(true)
        .NegativeAckRedeliveryDelay(TimeSpan.FromMilliseconds(500))
        .BatchReceivePolicy(new BatchReceivePolicy(100, -1L, TimeSpan.FromSeconds(10)))
        .SubscribeAsync();

var messages = await consumer.BatchReceiveAsync(CancellationToken);
var len = messages.Count; // len = ReceiverQueueSize(default: 1000)
```

**Comment by Lanayx** [comment]
Hi! BatchReceiveAsync is actually covered by a number of integration tests and I have never seen it failing. Can you please create an additional failing Integration test in PR so it would be possible to identify an issue and fix it.

**Comment by Myron5787** [comment]
I am unable to push the branch due to insufficient permissions.

**Comment by Lanayx** [comment]
It's probably your first PR :) You fork the repository, do you commit there and then create a PR from your repository to this one

**Comment by Myron5787** [comment]
Thank you for your explanation.
I have created the PR. Since I primarily work on C# projects, I would like to ask how the integration tests are generally conducted for this project.

**Comment by Lanayx** [comment]
https://github.com/fsprojects/pulsar-client-dotnet/pull/260 - I think you updated the right test, however the test didn't fail

**Comment by Lanayx** [comment]
If you know some specific failing conditions like more threads, more messages or anything else, I'd recommend to create a separate test rather than updating existing one

**Comment by Myron5787** [comment]
Currently, I am not aware of any additional triggering conditions, and I am not proficient in using F#. If I find any further triggering conditions later, I will report back.

It has been observed that in the .Net 8.0 C# environment, the data length returned by BatchReceiveAsync is equal to ReceiveQueueSize, rather than the maxNumMessages defined in BatchReceivePolicy.

Pulsar-broker image: apachepulsar/pulsar-all:2.10.2

**Comment by Lanayx** [comment]
I see, unfortunately without a failing test there is nothing to fix. I think it shouldn't be difficult to write additional test by copying existing one and modifying it. If you have any F#-related questions you can ask in [slack](https://fsharp.org/guides/slack/) or [discord](https://discord.com/invite/R6n7c54) or [telegram](https://t.me/fsharp_chat/26609), we have a friendly community

**Comment by CIRCULARKA** [comment]
+1 here, same problem - max limit of messages works only on first batch accept, then it allows messages more than specified amount. Is the behavior when message count limit is ignored due to compound messages size is lower then specified limit is working the way it expected to?

**Comment by Lanayx** [comment]
@CIRCULARKA no, message count limit is not expected to be ignored. You are welcome to create a repro (or even better - failing test) for the issue.

**Comment by adrian-apostolov** [comment]
+1 same problem..

<img width="657" height="189" alt="Image" src="https://github.com/user-attachments/assets/c3cfd447-cc83-41b0-b033-b28896d7b391" />

**Comment by Lanayx** [comment]
@adrian-apostolov Hi, same comments as above - without proper repro and/or relevant information about library version and configuration with debug logs there is nothing to do here.

**Comment by Ttpetrov** [comment]
+1 
Hello, I added a failing integration test and opened a pr =>[335](https://github.com/fsprojects/pulsar-client-dotnet/pull/335)

**Comment by Lanayx** [comment]
Fixed in 3.13.0
